### PR TITLE
Rest Apis to export/import with common code base

### PR DESCRIPTION
**Pull request by swakrish**
1)  Rest apis to export notebook as JSON
2)  Rest api to import notebook from JSON
https://issues.apache.org/jira/browse/ZEPPELIN-590

### CODE REVIEW

**Comment by swakrish** [comment]
screen shot 2016-01-08 at 4 33 31 pm
attaching screenshot for export JUnits

**Comment by swakrish** [comment]
screen shot 2016-01-08 at 4 35 15 pm
attaching screenshots for import Junit

**Comment by Leemoonsoo** [comment]
Thanks @swakrish for the contribution! 
Do you mind update `docs/rest-api/rest-notebook.md`, too? Changing this file updates zeppelin website.

**Comment by swakrish** [comment]
I added them

**Comment by bzz** [comment]
Thank you for improvement!

Reconciling the exisiting GUI import/export features to use this API may be a good idea too, otherwise we'll have different code doing the same job. What do you think?

Buy the way, could you also please check that your code formated accodring to the project styleguide?

**Comment by swakrish** [comment]
@bzz thanks for your comments. I did look at the UI functions, it doesnt have an import I guess and export is more or less similar. Let me know your thoughts

I also formatted it for google style.

**Comment by felixcheung** [comment]
I think maybe import should be a POST? In REST, PUT should be modifying an existing object?
http://restcookbook.com/HTTP%20Methods/put-vs-post/

**Comment by swakrish** [comment]
@felixcheung Yes makes sense, I changed it to http POST.

Please review and let me know your comments

**Comment by felixcheung** [comment]
Cool thanks.

This is the PR for notebook import #374
Agree with @bzz I think it makes sense if that is using this new REST API too?

**Comment by swakrish** [comment]
Does it make sense to have a separate PR for making UI import use REST API?

**Comment by corneadoug** [comment]
Right, it would make sense that both WebSocketAPI and REST API use the same function for Import.

**Comment by corneadoug** [comment]
Just to clarify, the UI should not use the REST API, the WebsocketAPI and the REST API should rely on the same code to do the import on the backend

**Comment by swakrish** [comment]
@corneadoug thanks for the comments, Just to clarify, I would be touchig the notebook.java to create a import/export method which can then be leveraged by Notebookserver.java for UI right?

Please correct me if im wrong

**Comment by corneadoug** [comment]
@swakrish Right exactly, that method would be used in both Websocket and REST API case :)

**Comment by swakrish** [comment]
@corneadoug I moved the code to common notebook.java. Please review and let me know. Thanks

**Comment by corneadoug** [comment]
Shouldn't the REST API send back an error response when something goes bad?

**Comment by swakrish** [comment]
@corneadoug 
it does, it throws IOException which results in Http 500 for import failed. This is similar to all other rest api method calls

I added a id not found for export too now, so it should return HTTP 500 too

**Comment by swakrish** [comment]
@corneadoug i donno why travis is failing. it looks like its failing on lots of PRs. Let me know if something else is missing for my commit too

**Comment by bzz** [comment]
There are couple of interesting things in logs:
 1.  At the very end:
 
   ```
9479K downloaded
[INFO] Expanding: /home/travis/build/apache/incubator-zeppelin/spark-dependencies/target/spark-dist/spark-1.1.1.tgz into /home/travis/build/apache/incubator-zeppelin/spark-dependencies/target/spark-dist
The log length has exceeded the limit of 4 Megabytes (this usually means that test suite is raising the same exception over and over).
The build has been terminated.
   ```

 2. Right before that, there are tests with strange exceptions

  ```
9:39:00,009 ERROR org.quartz.core.JobRunShell:211 - Job note.2BB4UKWE4 threw an unhandled Exception: 
java.lang.NullPointerException
	at org.apache.zeppelin.notebook.Notebook$CronJob.execute(Notebook.java:484)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
09:39:00,010 ERROR org.quartz.core.ErrorLogger:2425 - Job (note.2BB4UKWE4 threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: java.lang.NullPointerException]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: java.lang.NullPointerException
	at org.apache.zeppelin.notebook.Notebook$CronJob.execute(Notebook.java:484)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	... 1 more
09:39:00,001 ERROR org.quartz.core.JobRunShell:211 - Job note.2BB4UKWE4 threw an unhandled Exception: 
java.lang.NullPointerException
	at org.apache.zeppelin.notebook.Notebook$CronJob.execute(Notebook.java:484)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
09:39:00,010 ERROR org.quartz.core.ErrorLogger:2425 - Job (note.2BB4UKWE4 threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: java.lang.NullPointerException]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: java.lang.NullPointerException
	at org.apache.zeppelin.notebook.Notebook$CronJob.execute(Notebook.java:484)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	... 1 more
  ```

 3. another one, strange

  ```
  09:39:05,622 ERROR org.apache.zeppelin.rest.NotebookRestApi:341 - Exception in NotebookRestApi while moveParagraph 
java.lang.IndexOutOfBoundsException: paragraph size is 2 , index is 10
	at org.apache.zeppelin.notebook.Note.moveParagraph(Note.java:265)
	at org.apache.zeppelin.rest.NotebookRestApi.moveParagraph(NotebookRestApi.java:335)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:180)
  ```

 4. one more suspicious thing

  ```
09:39:57,491  INFO org.apache.zeppelin.interpreter.remote.RemoteInterpreterServer:551 - Exception in RemoteInterpreterServer while angularObjectAdd, nolock
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 5
	at com.google.gson.Gson.fromJson(Gson.java:802)
	at com.google.gson.Gson.fromJson(Gson.java:757)
	at com.google.gson.Gson.fromJson(Gson.java:706)
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreterServer.angularObjectAdd(RemoteInterpreterServer.java:546)
	at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Processor$angularObjectAdd.getResult(RemoteInterpreterService.java:1231)
	at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Processor$angularObjectAdd.getResult(RemoteInterpreterService.java:1216)
	at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)
	at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)
	at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:285)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 5
	at com.google.gson.stream.JsonReader.expect(JsonReader.java:339)
	at com.google.gson.stream.JsonReader.beginObject(JsonReader.java:322)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:184)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:146)
	at com.google.gson.Gson.fromJson(Gson.java:791)
	... 11 more
  ```

Each of them looks very strange to me, never seen any of them before and each worth further investigation.

**Comment by swakrish** [comment]
@Leemoonsoo issue 2 has been happening before too
3 & 4 seems to have started happening from sometime back. Do we need a seperate JIRA/PR for those?

**Comment by swakrish** [comment]
@bzz 
are any related to this PR ? I ran my unit tests standalone and seems to be working fine

**Comment by bzz** [comment]
@swakrish I assume that you, as an author, should know better. 

Usually the mindset we try to share here is - CI must be green for anything to be merged.  If it's not - then dedicated effort worth putting is - try to fix it i.e by collaborating with more experienced members, asking for help, etc. 

If you have some spare cycles you could spend them creating a separate issues, investigating these failures and pinging the people who touched particular code before - that would be very appreciated and of the great help to the project. 

Otherwise best bet is to wait until somebody takes care of this one day, so you can rebase on top of that work, but this might take some time.

**Comment by swakrish** [comment]
@bzz thanks for your comments, i just wanted to confirm since when i run test suite it doesnt break any of my changes :)

I will take a look at those failures you mentioned to see why those are happening. Again those doesnt seem to happen in local builds

**Comment by swakrish** [comment]
@bzz , point 3 is negative test case, so i think its fine, the unit test dont fail here

**Comment by bzz** [comment]
Great! 
One more thing - @swakrish  @corneadoug  could you guys please point me to the place where REST API and WS API for importing\exporting notebooks re-use the same code path?

**Comment by swakrish** [comment]
@bzz Please take a look at notebook.java, this is accessible both from REST and Websocket,
We need to modify notebookserver.java to use the common method for UI

File: docs/rest-api/rest-notebook.md [hunk 1]
```diff
@@ -773,3 +773,112 @@ limitations under the License.
...
+  
+  <table class="table-configuration">
+    <col width="200">
+    <tr>
+      <th>Export notebook</th>
+      <th></th>
+    </tr>
+    <tr>
+      <td>Description</td>
+      <td>This ```PUT``` method imports a notebook from the notebook JSON
```
Comment (felixcheung) [inline docs/rest-api/rest-notebook.md]: oh we might have missed updating this since it's been changed to POST...

Comment (swakrish) [inline docs/rest-api/rest-notebook.md]: @felixcheung thanks for the comments, i changed the doc, please review

**Comment by swakrish** [comment]
Please review this

**Comment by bzz** [comment]
Sorry for a delay. Will review this today.

**Comment by swakrish** [comment]
thanks @bzz

**Comment by bzz** [comment]
@swakrish sorry, may be I did not make it clear last time, but having 2 code paths for same function (for importing\exporting notebooks) in different APIs (WS\REST) does not look as a good idea to me.

> @bzz Please take a look at notebook.java, this is accessible both from REST and Websocket
I did, and both of them are called exactly once, in the `NotebookRestApi.java`

This brings us back to the question asked before:
>One more thing - @swakrish @corneadoug could you guys please point me to the place where REST API and WS API for importing\exporting notebooks re-use the same code path?

As far as I can tell such place does not exist, right now. This makes concerns, expressed by me and @corneadoug in this PR not addressed. Please let me know if I'm missing something here.

If you have a plan on how to converge these API's code paths (yours and one from the [Websocket](https://github.com/apache/incubator-zeppelin/blob/master/zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java#L136)) - why not to do it in this PR?  I think that way it would be a even greater contribution, reducing number of possible bugs, etc. What do you think?

**Comment by swakrish** [comment]
@bzz Yes i was thinking of opening a different PR for that since this address just REST calls. I have a tested code base for modifying web socket too
Let me know if creating a separate PR doesnt make sense ?
snippet:
protected Note importNote(NotebookSocket conn, Notebook notebook, Message fromMessage)
      throws IOException {
    Note note = null;
    if (fromMessage != null) {
      String noteName = (String) ((Map) fromMessage.get("notebook")).get("name");
      String noteJson = gson.toJson(fromMessage.get("notebook"));
      note = notebook.importNote(noteJson, noteName);
      note.persist();
      broadcastNote(note);
      broadcastNoteList();
    }
    return note;
  }

**Comment by swakrish** [comment]
@bzz I merged the websocket import to use the common code.So now both REST/UI import uses same code.
Export is just getting the json notation of the note, so I feeel we dont need to  change the code base to go through same path. Let me know if you still feel we need to create a websocket call

**Comment by swakrish** [comment]
@bzz can you review this?

**Comment by bzz** [comment]
@swakrish import code looks great.

Just to double-check, I think it's good to ask original author of `ImportNote` from #374 to review this.
@prabhjyotsingh if you have a few spare cycles - could you take a look please?

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java [hunk 2]
```diff
@@ -150,6 +154,57 @@ public Note createNote(List<String> interpreterIds) throws IOException {
...
+    Note note = getNote(noteId);
+    if (note == null) {
+      throw new IllegalArgumentException(noteId + "not found");
+    }
+    return gson.toJson(note);
+  }
+
+  /**
+   * import JSON as a new note.
+   * @param sourceJSON - the note JSON to import
```
Comment (prabhjyotsingh) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java]: sourceJSON -> sourceJson

**Comment by prabhjyotsingh** [comment]
Test this PR, both newer and existing functionality works fine.

**Comment by swakrish** [comment]
@prabhjyotsingh , thanks for checking this so quickly and your comments.
I modified the comment

**Comment by swakrish** [comment]
@prabhjyotsingh @bzz , please review

**Comment by swakrish** [comment]
@prabhjyotsingh  @bzz  , please review

**Comment by swakrish** [comment]
is there any other change needed on this PR?

**Comment by bzz** [comment]
@prabhjyotsingh thanks for prompt reply!

Looks great to me, will merge if there is no more discussion.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java [hunk 3]
```diff
@@ -150,6 +154,57 @@ public Note createNote(List<String> interpreterIds) throws IOException {
...
+   * @return Note JSON
+   * @throws IOException, IllegalArgumentException
+   */
+  public String exportNote(String noteId) throws IOException, IllegalArgumentException {
+    GsonBuilder gsonBuilder = new GsonBuilder();
+    gsonBuilder.setPrettyPrinting();
+    Gson gson = gsonBuilder.create();
+    Note note = getNote(noteId);
+    if (note == null) {
+      throw new IllegalArgumentException(noteId + "not found");
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java]: suggest `noteId + " not found"` for extra space

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java [hunk 4]
```diff
@@ -150,6 +154,57 @@ public Note createNote(List<String> interpreterIds) throws IOException {
...
+      else
+        newNote.setName(oldNote.getName());
+      List<Paragraph> paragraphs = oldNote.getParagraphs();
+      for (Paragraph p : paragraphs) {
+        newNote.addCloneParagraph(p);
+      }
+
+      newNote.persist();
+    } catch (IOException e) {
+      throw new IOException(e);
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java]: sorry coming in late - could we log here
```
             LOG.error("Message", e);
```
alternatively, let's remove the `catch {} throw{}` if we end up wrapping with the same type

Comment (swakrish) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java]: @felixcheung , we need the catch to log the exception right? so i will just add logger?
   logger.error(e.toString(), e);

**Comment by swakrish** [comment]
@felixcheung pleasereview, i think travis is stuck :)

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java (hunk 4 above)
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java]: right, this is great
```
   } catch (IOException e) {
       logger.error(e.toString(), e);
       throw e;
     }
```

**Comment by felixcheung** [comment]
looks good

**Comment by swakrish** [comment]
@felixcheung thanks for reviewing, looks like travis finally passed

**Comment by bzz** [comment]
@swakrish could you assign a JIRA issues to yourself please? Or let me know your JIRA id and I can do that for you.

**Pull request closed without merging**